packages/core/src/codewhisperer/nextEditPrediction/PredictionTracker.ts

```diff
@@ -0,0 +1,394 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import * as path from 'path'
+import fs from '../../shared/fs/fs'
+import { getLogger } from '../../shared/logger/logger'
+import * as diffGenerator from './diffContextGenerator'
+import * as codewhispererClient from '../client/codewhisperer'
+import { predictionTrackerDefaultConfig } from '../models/constants'
+
+const snapshotDirName = 'AmazonQ-file-snapshots'
+const snapshotFileSuffix = '.nep-snapshot'
+
+// defaul values are stored in codewhisperer/model/constants
+export interface FileTrackerConfig {
+    maxFiles: number
+    maxStorageSizeKb: number
+    debounceIntervalMs: number
+    maxAgeMs: number
+    maxSupplementalContext: number
+}
+
+/**
+ * Represents a snapshot of a file at a specific point in time
+ */
+export interface FileSnapshot {
+    filePath: string
+    size: number
+    timestamp: number
+    storageKey: string
+}
+
+export class PredictionTracker {
+    private snapshots: Map<string, FileSnapshot[]> = new Map()
+    readonly config: FileTrackerConfig
+    private storageSize: number = 0
+    private storagePath: string
+
+    constructor(extensionContext: vscode.ExtensionContext, config?: Partial<FileTrackerConfig>) {
+        this.config = {
+            ...predictionTrackerDefaultConfig,
+            ...config,
+        }
+
+        this.storagePath = extensionContext.storageUri?.fsPath as string
+
+        void this.ensureStorageDirectoryExists()
+        void this.loadSnapshotsFromStorage()
+    }
+
+    public processEdit(document: vscode.TextDocument, previousContent: string): void {
+        const filePath = document.uri.fsPath
+        getLogger().debug(`Processing edit for file: ${filePath}`)
+
+        if (!this.storagePath || !document.uri.scheme.startsWith('file')) {
+            return
+        }
+
+        void this.takeSnapshot(filePath, previousContent)
+    }
+
+    /**
+     * Takes a snapshot with provided previous content
+     * @param filePath The path with of document
+     * @param previousContent It's content before the edit
+     */
+    private async takeSnapshot(filePath: string, previousContent: string): Promise<void> {
+        const content = previousContent
+        const size = Buffer.byteLength(content, 'utf8')
+
+        const timestamp = Date.now()
+        const storageKey = `${filePath.replace(/\//g, '__')}-${timestamp}`
+
+        const snapshot: FileSnapshot = {
+            filePath,
+            size,
+            timestamp,
+            storageKey,
+        }
+
+        // Get existing snapshots for this file
+        const fileSnapshots = this.snapshots.get(filePath) || []
+
+        // Check if we should add a new snapshot given the debounce
+        const shouldAddSnapshot =
```

Comment:
Should we check this condition as early as possible in this function and return if false. Then you won't need to do as much unneeded setup.